### PR TITLE
More work towards feature parity of 1x

### PR DESCRIPTION
**Pull request by matanlurey**
Added:

* `Allocator` (prefixing)
* `Directive` and `File`

### CODE REVIEW

**Comment by matanlurey** [comment]
Ping @alorenzen since you're back - this is probably enough to roll I imagine.

File: lib/src/emitter.dart [hunk 1]
```diff
@@ -2,18 +2,30 @@
...
+import 'specs/file.dart';
 import 'specs/method.dart';
 import 'specs/reference.dart';
 import 'specs/type_reference.dart';
 import 'visitors.dart';
 
-class DartEmitter extends GeneralizingSpecVisitor<StringSink> {
-  const DartEmitter();
+class DartEmitter implements SpecVisitor<StringSink> {
+  final Allocator _allocator;
```
Comment (alorenzen) [inline lib/src/emitter.dart]: I don't think _allocator.imports is ever called. So how would you actually print out the imports in a file?

File: test/specs/file_test.dart [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import 'package:code_builder/code_builder.dart';
+import 'package:test/test.dart';
+
+void main() {
+  group('File', () {
+    const $linkedHashMap = const Reference('LinkedHashMap', 'dart:collection');
+    test('should emit a source file', () {
+      expect(
+        new File((b) => b
+          ..directives.add(new Directive.import('dart:collection'))
```
Comment (alorenzen) [inline test/specs/file_test.dart]: Is the assumption that you always have to add the imports manually? Can you add a test which uses allocator.imports instead?

Also, maybe a test case for prefixing imports?

File: lib/src/allocator.dart [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  /// style and instead takes a conservative approach of prefixing _every_
+  /// import except references to `dart:core` (which are considered always
+  /// imported).
+  factory Allocator.simplePrefixing() = _PrefixedAllocator;
+
+  /// Returns a reference string given a [reference] object.
+  ///
+  /// For example, a no-op implementation:
+  /// ```dart
+  /// allocate(const Reference('List', 'dart:core')); // Outputs 'List'.
```
Comment (matanlurey) [inline lib/src/allocator.dart]: Done.

File: lib/src/allocator.dart [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  /// ```
+  ///
+  /// Where-as an implementation that prefixes imports might output:
+  /// ```dart
+  /// allocate(const Reference('Foo', 'package:foo')); // Outputs '_i1.Foo'.
```
Comment (matanlurey) [inline lib/src/allocator.dart]: Done.

File: lib/src/emitter.dart (hunk 1 above)
Comment (matanlurey) [inline lib/src/emitter.dart]: Thanks, good catch. Added to `visitFile` and updated `file_test`.

File: test/specs/file_test.dart (hunk 2 above)
Comment (matanlurey) [inline test/specs/file_test.dart]: Fixed. Added test cases.

**Pull request merged**